### DISCUSSION TITLE
在circom中，约束的类型如何理解？

### DISCUSSION THREAD

**Discussion by timimm** [zkp-co-learning/zkp-co-learn#41] [Q&A（提问题在隔壁~）] [answered]
# 在circom中，约束的类型如何理解？
## 0x01 背景
测试模逆运算时，代码如下：
```circom
pragma circom 2.1.2;

template Example () {
    signal input a;
    signal output b;

    b <-- 1/a; 
}

component main = Example();

/* INPUT = {
    "a" : "5"
} */
```
代码可以正确运行，并得出答案。
但是对`b<--1/a`，施加约束变为`b<==1/a`就会报错
报错信息为：`Non quadratic constraints are not allowed!`.
修改代码如下，又可正常运行：
```circom
pragma circom 2.1.2;

template Example () {
    signal input a;
    signal output b;

    b <-- 1/a;
    b * a === 1;
    
}

component main = Example();

/* INPUT = {
    "a" : "5"
} */
```
## 0x02 问题
- circom对约束的类型是怎样规定的？所谓二次约束是啥意思？

**Comment by izayl** [comment, marked as answer]
https://docs.circom.io/circom-language/constraint-generation/
> Constant values: only a constant value is allowed.
Linear expression: an expression where only addition is used. It can also be written using multiplication of variables by constants. For instance, the expression 2*x + 3*y + 2 is allowed, as it is equivalent to x + x + y + y + y + 2.
Quadratic expression: it is obtained by allowing a multiplication between two linear expressions and addition of a linear expression: A*B - C, where A, B and C are linear expressions. For instance, (2*x + 3*y + 2) * (x+y) + 6*x + y – 2.
Non quadratic expressions: any arithmetic expression which is not of the previous kind.

**Reply by izayl** [reply]
> Other arithmetic expressions beyond quadratic or using other arithmetic operators like division or power are not allowed as constraints.

**Comment by LowLowCoder-Mel** [comment]
0x1
上面的代码中，b <-- 1/a 表达式是不符合二次约束的。在 Circom 中，二次约束是指所有信号的次数（degree）都小于等于 2 的约束，或者存在一些信号的次数为 2，其他信号的次数为 1 的约束。这是为了确保约束满足可满足性问题（SAT）求解器的性质，以便实现有效的约束求解。

在这个例子中，表达式 b <-- 1/a 涉及到信号 a 的次数为 -1，不符合双二次约束的要求。因此，这段代码段传递给 Circom 编译器，则会报告 Non quadratic constraints are not allowed 的编译错误。

要修复这个错误，可以使用其他约束方式，例如将 b 定义为 a 的倒数的乘积，即 a * b == 1，这样就满足了双二次约束的要求。修改后的代码如下：
template Example () {
    signal input a;
    signal output b;
    a * b == 1; 
}

0x2
在 Circom 中，二次约束是指所有信号的次数（degree）都小于等于 2 的约束，或者存在一些信号的次数为 2，其他信号的次数为 1 的约束。
以上为个人理解，如果错误望大佬指正。

**Reply by timimm** [reply]
明白了！
